Run lint and upload results to Github Code Scanning

wzieba opened the pull request:
Closes #118 

### Description

This PR integrates Android Lint with GitHub Code Scanning support.

### Testing Steps

Green light from CI is just fine.

wzieba committed:
Run lint and upload results to Github Code Scanning

wzieba committed:
Update CI toolkit

wzieba committed:
Apply ci toolkit to the lint step

wzieba committed:
Update CI toolkit

wzieba committed:
Use release variant for lint

wzieba committed:
Update comment

wzieba committed:
Add lint.xml configuration file

wzieba committed:
Promote `UnusedResources` to error

AdamGrzybkowski commented on .buildkite/commands/lint.sh:
```diff
@@ -0,0 +1,10 @@
+
+echo "--- 🧹 Linting"
+# Run lint on the app module first to ensure that the SARIF file is always generated.
+./gradlew :gravatar-quickeditor:lintRelease
```
Technically `:quickeditor` is not an app module. Will the lint check also run against the `:ui` module?

hamorillo replied:
> Will the lint check also run against the :ui module?

I've made a quick test, and I would say we need to run lint for each module.

wzieba replied:
Because `quickeditor` uses `ui` module, the lint will also check the `ui` module (see: `checkDependencies = true`)

I configured it this way, because I want us to have a single lint report that we can send to GitHub Code Scanning registry. If we uploaded per-module reports, they would override each other.

AdamGrzybkowski replied:
I guess that should work, we just need to keep in mind that this has to be updated if we have new modules.

How would that setup look if we had a new module that depends on the `:ui` but not on `:quickeditor`? 

Could we maybe run the lint against the demo app, but ignore errors/warnings from the demo app module?

hamorillo replied:
> Because quickeditor uses ui module, the lint will also check the ui module (see: checkDependencies = true)

You are right. I've rechecked it, and it works as you described. Thanks!

wzieba replied:
> Could we maybe run the lint against the demo app, but ignore errors/warnings from the demo app module?

Neat idea @AdamGrzybkowski , thanks! I think it's doable, with a little tweak in `lint.xml`: [ece18ea](https://github.com/Automattic/Gravatar-SDK-Android/pull/315/commits/ece18eaede2aaab46c95721b10130c41cd9c7e99)

When working on this, I've realised that I had misconfigured `lint.xml` previously. The updated configuration is now correct, but it reports errors. ~What should we do with them?~ I proposed fixes here

AdamGrzybkowski replied:
I missed some uCrop related code in the `:ui` module that was making the Lint to fail (already useful :D). I pushed the changes.


EDIT: There was also some leftover `java.time` code 🙃

wzieba committed:
Move lint configuration to `demo-app`

And ignore all issues in `demo-app`. Assuming that all sub-modules will be always used in `demo-app`, this setup ensures a unified lint report for SDK artifacts.

wzieba committed:
Ignore `UnusedResources` on `available_languages`

wzieba committed:
Remove redundant lint configuration

wzieba commented on gravatar-quickeditor/src/main/res/values/available_languages.xml:
```diff
@@ -1,7 +1,7 @@
 <!--Warning: Auto-generated file, do not edit.-->
 <resources xmlns:tools="http://schemas.android.com/tools">
-  <string-array name="available_languages" translatable="false" tools:ignore="InconsistentArrays">
+  <string-array name="available_languages" translatable="false" tools:ignore="InconsistentArrays, UnusedResources">
```
Is this ok @iangmaia ? I don't know how GlotPress integration works: is this file automatically generated, or is it fine to add configuration to it like this?

iangmaia replied:
It was originally auto-generated, but I think this will be fine (as in: I don't think this will be overwritten, only the locale list should be).

wzieba committed:
Merge branch 'trunk' into integrate_lint

wzieba committed:
Bump ci toolkit to 3.7.0

wzieba committed:
Remove unnecessary comment

AdamGrzybkowski committed:
Remove leftover uCrop related code from the :ui module

AdamGrzybkowski committed:
Remove java.time code

AdamGrzybkowski reviewed (approved):
Thanks @wzieba!

wzieba commented:
Thanks for the fixes @AdamGrzybkowski - it's great that lint caught e.g. 1b9a5f1!
